### ISSUE TITLE
Implement ValueProvider, FactoryProvider like it is done in angular 2

### ISSUE THREAD

**Issue by RobinBuschmann** [RobinBuschmann/ts-di#1]
Add possibility to inject a value or a factory(function). See [this](https://angular.io/docs/ts/latest/guide/dependency-injection.html#!#non-class-dependencies) for further information.

**Comment by KostyaTretyak** [comment]
@RobinBuschmann, if we do this, we'll have full Angular DI v2.4.1:

```bash
git clone git@github.com:angular/angular.git
cd angular
git checkout 2.4.x
cd ..
mkdir -p myproject/src
cd myproject/src
mkdir facade util reflection
cp -r ../../angular/modules/@angular/core/src/di .
cp ../../angular/modules/@angular/core/src/di.ts .
cp ../../angular/modules/@angular/core/src/facade/lang.ts facade
cp ../../angular/modules/@angular/core/src/facade/errors.ts facade
cp ../../angular/modules/@angular/core/src/type.ts .
cp ../../angular/modules/@angular/core/src/util/decorators.ts util
cp ../../angular/modules/@angular/core/src/reflection/reflection.ts reflection
cp ../../angular/modules/@angular/core/src/reflection/reflection_capabilities.ts reflection
cp ../../angular/modules/@angular/core/src/reflection/reflector.ts reflection
cp ../../angular/modules/@angular/core/src/reflection/platform_reflection_capabilities.ts reflection
cp ../../angular/modules/@angular/core/src/reflection/reflector_reader.ts reflection
cp ../../angular/modules/@angular/core/src/reflection/types.ts reflection
```

And then we can do this:

```ts
import 'reflect-metadata';

import {
  Injectable
  ,Inject
  ,ReflectiveInjector
  ,OpaqueToken
} from './src/di';


interface Engine
{
  one: number;
};

const ENGINE: Engine =
{
  one: 1
};

let asToken = new OpaqueToken('some Tok');

@Injectable()
class Car
{
  constructor( @Inject(asToken) public engine: Engine ){}

  start() {}
}

var injector = ReflectiveInjector.resolveAndCreate([Car, {provide: asToken, useValue: ENGINE }])
var car = injector.get(Car);

/**
 * Expect `car` to be instance of class `Car`
 * and `car.engine` to be instance of class `MockEngine`
 */

console.log(car);
```

It seems that lacks only mechanism for add providers in this array: `ReflectiveInjector.resolveAndCreate([])`.

For Angular DI, providers must be added separately. At first glance it is inconvenient, but it just allows substitution providers and simplifies testing.

**Comment by RobinBuschmann** [comment]
Probably the `useClass`/`useValue`approach of angular 2 is not necessary. Since a Token - in ng2 it is called OpaqueToken - is necessary to provide a primitive value or an object literal, currently I see no  disadvantage in using a factory instead. A factory can provide everything you want, because it is a simple function.

So, in angular 2 it is working like this:

```typescript
import { OpaqueToken } from '@angular/core';

const ENV = new OpaqueToken('env.var');
const env = 'development';

@Component({
  providers: [{ provide: ENV, useValue: env }]
})
class SomeComponent {

  constructor(@Inject(ENV) env: string) {

  }
}
```
Due to the creation of an OpaqueToken there is some extra work todo anyway. 

But what about:

```typescript
const env = 'development';
const envFactory = () => env;

@Inject
class SomeService {

  constructor(@useFactory(envFactory) env: string) {

  }
}
```

Regarding testing we could also simplify ng2 approach:

```typescript

const specialFactory = () => 'special value';
const injector = new Injector([{token: envFactory, use: specialFactory}]);


```

This looks much simpler from my point of view and we can omit the OpaqueToken. What do you think @KostyaTretyak?

**Comment by RobinBuschmann** [comment]
@KostyaTretyak So, you're thinking about forking the di part of ng2 and using this instead?

**Comment by KostyaTretyak** [comment]
@RobinBuschmann, I will learn it, but not sure if I can rework it for the purposes of Node's backend.

**Comment by RobinBuschmann** [comment]
@KostyaTretyak What do you think about my suggestion? https://github.com/RobinBuschmann/ts-di/issues/1#issuecomment-275094123

**Comment by KostyaTretyak** [comment]
@RobinBuschmann, I think that Angular team are more skilled than you and I put together. Maybe we not see how their approach is cool, but I trust them.

**Comment by KostyaTretyak** [comment]
By the way, ng's DI also [has factory](https://github.com/angular/angular/blob/77ee27c59e878f6fb5c040d1421266716d811b0f/modules/%40angular/core/test/di/reflective_injector_spec.ts#L166). And this not always good, because result from factory not caching (I think), and DI must always go a bit long road.

**Comment by RobinBuschmann** [comment]
Caching is a good point, but [angular/di.js](https://github.com/angular/di.js) already do this for us. It also has a factory implementation ([see](https://github.com/angular/di.js/blob/master/src/providers.js#L131)) with(!) caching ([see](https://github.com/angular/di.js/blob/master/src/injector.js#L279)). But I couldn't found out how to provide a factory when using angular/di.js

I'm excited to see how you get it done :)

**Comment by KostyaTretyak** [comment]
=) I just will try to use Angular DI:

```ts
import * as di from '@angular/core/src/di';
```

I used to think that the DI scattered throughout Angular, and it can not be used as a standalone module. But now I see it's possible.

**Comment by RobinBuschmann** [comment]
Is there a way to see your progress on this? Did you pushed the repo to github?

**Comment by KostyaTretyak** [comment]
You talk about progress `ts-di` module? This module not work in progres, for now.

Now I work for [restify-ts](https://github.com/restify-ts), and try implements Angular DI for node's backend.

**Comment by KostyaTretyak** [comment]
See [this](https://github.com/restify-ts/core/blob/master/lib/server.ts#L108) - it's old try implements `ts-di`.

**Comment by RobinBuschmann** [comment]
With progress I mean, how far you are with your extraction of the angular 2 di and whether it is possible or not to see what you have done yet :)

**Comment by KostyaTretyak** [comment]
All my work about "extraction" Angular DI:

```ts
import * as di from '@angular/core/src/di';
```

**Comment by RobinBuschmann** [comment]
so, its implemented: 9fea85e17414993cc9d9e41efb750e01e7c7edad
